File: frontend/src/lib/legacyFallback.js
const STORE_KEY = "__MES_LEGACY_FALLBACK_HITS__";

const normalizeText = (value) => String(value ?? "").trim();
const SAFE_DETAIL_KEYS = new Set([
  "domain",
  "reason",
  "room",
  "source",
  "status",
  "targetIsFallback",
  "targetType",
]);

const createStore = () => ({
  hits: new Map(),
  sequence: 0,
});

const getStore = () => {
  if (!globalThis[STORE_KEY]) {
    globalThis[STORE_KEY] = createStore();
  }
  return globalThis[STORE_KEY];
};

export function recordLegacyFallbackHit(id, detail = {}) {
  const normalizedId = normalizeText(id);
  if (!normalizedId) {
    return null;
  }
  const store = getStore();
  const previous = store.hits.get(normalizedId) || { count: 0, id: normalizedId, lastDetail: null, lastSequence: 0 };
  const safeDetail = Object.fromEntries(
    Object.entries(detail && typeof detail === "object" ? detail : {})
      .filter(([key]) => SAFE_DETAIL_KEYS.has(key))
      .map(([key, value]) => [key, normalizeText(value)]),
  );
  const next = {
    ...previous,
    count: previous.count + 1,
    lastDetail: safeDetail,
    lastSequence: store.sequence + 1,
  };
  store.sequence += 1;
  store.hits.set(normalizedId, next);
  return next;
}

export function getLegacyFallbackHits() {
  return Array.from(getStore().hits.values())
    .map((hit) => ({ ...hit, lastDetail: hit.lastDetail ? { ...hit.lastDetail } : hit.lastDetail }))
    .sort((left, right) => left.id.localeCompare(right.id));
}

export function resetLegacyFallbackHits() {
  globalThis[STORE_KEY] = createStore();
}

export function legacyFallbackHitCount(id) {
  const normalizedId = normalizeText(id);
  if (!normalizedId) {
    return 0;
  }
  return getStore().hits.get(normalizedId)?.count || 0;
}

File: app/core/legacy_fallback.py
from __future__ import annotations

from collections import Counter
import logging
from typing import Any

logger = logging.getLogger(__name__)
_hits: Counter[str] = Counter()
_last_details: dict[str, dict[str, Any]] = {}
_SAFE_DETAIL_KEYS = {
    "domain",
    "reason",
    "room",
    "source",
    "status",
    "target_is_fallback",
    "target_type",
}


def record_legacy_fallback_hit(fallback_id: str, **detail: Any) -> None:
    normalized_id = str(fallback_id or "").strip()
    if not normalized_id:
        return
    safe_detail = {key: str(value) for key, value in detail.items() if key in _SAFE_DETAIL_KEYS}
    _hits[normalized_id] += 1
    _last_details[normalized_id] = safe_detail
    logger.info(
        "legacy fallback hit",
        extra={
            "fallback_id": normalized_id,
            "legacy_fallback": _last_details[normalized_id],
            "legacy_fallback_hit": True,
        },
    )


def get_legacy_fallback_hits() -> list[dict[str, Any]]:
    return [
        {
            "count": count,
            "id": fallback_id,
            "last_detail": dict(_last_details.get(fallback_id, {})),
        }
        for fallback_id, count in sorted(_hits.items())
    ]


def reset_legacy_fallback_hits() -> None:
    _hits.clear()
    _last_details.clear()
